### PR TITLE
Added notification for mismatched g2.dat entries size

### PR DESCRIPTION
**Pull request by jazzysoggy**
#16662

Added a value named EXPECTED_SPRITES in StringIds.h that holds the value of SPR_G2_END - SPR_G2_BEGIN

Added a check when loading g2.dat to see if _g2.header.num_entries is equal to the EXPECTED_SPRITES value.

If it isn't, a error will be logged and a notification will appear displaying "Mismatched g2.dat size".

### CODE REVIEW

File: src/openrct2/drawing/Drawing.Sprite.cpp [hunk 1]
```diff
@@ -282,6 +283,16 @@ bool gfx_load_g2()
         // Read element data
         _g2.data = fs.ReadArray<uint8_t>(_g2.header.total_size);
 
+        if (_g2.header.num_entries != EXPECTED_SPRITES)
+        {
+            log_error("Mismatched g2.dat size");
+            if (!gOpenRCT2Headless)
+            {
+                auto uiContext = GetContext()->GetUiContext();
+                uiContext->ShowMessageBox("Mismatched g2.dat size");
```
Comment (duncanspumpkin) [inline src/openrct2/drawing/Drawing.Sprite.cpp:292]: I feel this could be further embellished perhaps stating that files incorrectly installed.

Comment (janisozaur) [reply src/openrct2/drawing/Drawing.Sprite.cpp:292]: And provide counts received and expected. If possible, also path to g2.dat.

File: src/openrct2/localisation/StringIds.h [hunk 2]
```diff
@@ -3877,6 +3878,7 @@ enum : uint16_t
 
     STR_MAINTAIN_SQUARE_MAP_TOOLTIP = 6498,
 
+    EXPECTED_SPRITES = SPR_G2_END - SPR_G2_BEGIN,
```
[the remark this answers is not in the document]
Comment (jazzysoggy) [reply src/openrct2/localisation/StringIds.h]: Renamed variable to G2_SPRITE_COUNT and moved it to sprites.h

**Comment by karst** [comment]
Shouldn't this message have localization or is it unnecessary in this case?

**Comment by jazzysoggy** [comment]
One message for headless clients, one message for GUI clients?

File: src/openrct2/drawing/Drawing.Sprite.cpp [hunk 3]
```diff
@@ -282,6 +282,20 @@ bool gfx_load_g2()
         // Read element data
         _g2.data = fs.ReadArray<uint8_t>(_g2.header.total_size);
 
+        if (_g2.header.num_entries != G2_SPRITE_COUNT)
+        {
+            std::string errorMessage = "Mismatched g2.dat size.\nExpected: " + std::to_string(G2_SPRITE_COUNT) + "\nActual: "
+                + std::to_string(_g2.header.num_entries) + "\ng2.dat may be installed imporperly.\nPath to g2.dat: " + path;
```
Comment (janisozaur) [inline src/openrct2/drawing/Drawing.Sprite.cpp:288]:
```suggestion
                + std::to_string(_g2.header.num_entries) + "\ng2.dat may be installed improperly.\nPath to g2.dat: " + path;
```

File: src/openrct2/drawing/Drawing.Sprite.cpp [hunk 4]
```diff
@@ -282,6 +282,20 @@ bool gfx_load_g2()
...
+
+            log_error(errorMessage.c_str());
+
+            if (!gOpenRCT2Headless)
```
Comment (janisozaur) [inline src/openrct2/drawing/Drawing.Sprite.cpp:292]: I wonder if it is necessary, or if the same condition can be expressed with `uiContext != nullptr`? @IntelOrca

**Comment by janisozaur** [comment]
@jazzysoggy thanks so much, it's been long overdue! And for iterating on this so quickly

File: src/openrct2/drawing/Drawing.Sprite.cpp [hunk 5]
```diff
@@ -282,6 +282,20 @@ bool gfx_load_g2()
         // Read element data
         _g2.data = fs.ReadArray<uint8_t>(_g2.header.total_size);
 
+        if (_g2.header.num_entries != G2_SPRITE_COUNT)
+        {
+            std::string errorMessage = "Mismatched g2.dat size.\nExpected: " + std::to_string(G2_SPRITE_COUNT) + "\nActual: "
```
Comment (janisozaur) [inline src/openrct2/drawing/Drawing.Sprite.cpp:287]: The message could maybe mention the user is about to experience graphical glitches due to this

Comment (jazzysoggy) [reply src/openrct2/drawing/Drawing.Sprite.cpp:287]: Resolve by adding prompt that appears after first prompt detailing issue and how to fix.

File: src/openrct2/localisation/StringIds.h [hunk 6]
```diff
@@ -10,6 +10,7 @@
 #pragma once
 
 #include "../common.h"
+#include "../sprites.h"
```
Comment (tupaschoal) [inline src/openrct2/localisation/StringIds.h:13]: Why is this needed?

**Review by tupaschoal — COMMENTED** [review]
Please add a changelog entry for this

File: src/openrct2/localisation/StringIds.h (hunk 6 above)
Comment (Broxzier) [reply src/openrct2/localisation/StringIds.h:13]: It's a leftover from when `EXPECTED_COUNT` was defined in the string array, now renamed and moved to the sprite enum. This include should be removed.

Comment (jazzysoggy) [reply src/openrct2/localisation/StringIds.h]: Removed the redundant include

**Comment by jazzysoggy** [comment]
Could someone review this?
